Add documentation for Unhealthy Pod Eviction Policy for PDBs (#37768)
kubernetes/website: content/en/docs/tasks/run-application/configure-pdb.md

  observedGeneration: 1
```

## Unhealthy Pod Eviction Policy

{{< feature-state for_k8s_version="v1.26" state="alpha" >}}

{{< note >}}
In order to use this behavior, you must enable the `PDBUnhealthyPodEvictionPolicy`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
on the [API server](/docs/reference/command-line-tools-reference/kube-apiserver/).
{{< /note >}}

> PodDisruptionBudget guarding an application ensures that `status.currentHealthy` number of pods

Reviewer: ```suggestion
PodDisruptionBudget guarding an application ensures that `.status.currentHealthy` number of pods
```

> does not fall under `status.desiredHealthy` number by not allowing eviction of healthy pods.

Reviewer: ```suggestion
does not fall under `.status.desiredHealthy` number by not allowing eviction of healthy pods.
```

By using `.spec.unhealthyPodEvictionPolicy`, you can also define the criteria when unhealthy pods 
should be considered for eviction. The default behavior when no policy is specified corresponds 
to the `IfHealthyBudget` policy.

{{< note >}}

> Current implementation considers healthy pods, as pods that have `status.conditions` item with `type="Ready"` and `status="True"`.

Reviewer: ```suggestion
Current implementation considers healthy pods, as pods that have `.status.conditions` item with `type="Ready"` and `status="True"`.
```

> {{< /note >}}

Author: Do we want to define this in the doc? It is not mentioned anywhere else now.

Reviewer: Yeah, probably a separate section which you could use in other places as links (see my other comment).

Author: created a new section under PDB status called `Healthiness of a Pod`

Policies:

`IfHealthyBudget`

> : Running pods (`status.phase="Running"`), but not yet healthy can be evicted only if the guarded application is not 

Reviewer: ```suggestion
: Running pods (`.status.phase="Running"`), but not yet healthy can be evicted only if the guarded application is not 
```
and in all the remaining instances as well.

Author: thx, hopefully fixed all the dots

disrupted (`status.currentHealthy` is at least equal to `status.desiredHealthy`).

> : This policy ensures that running pods of an already disrupted application have the best change to become healthy. 

Reviewer: ```suggestion
: This policy ensures that running pods of an already disrupted application have the best chance to become healthy. 
```

This has negative implications for draining nodes, which can be blocked by misbehaving applications that are guarded by a PDB.
More specifically applications with pods in `CrashLoopBackOff` state (due to a bug or misconfiguration),
or pods that are just failing to report the `Ready` condition.

`AlwaysAllow`
: Running pods (`status.phase="Running"`), but not yet healthy are considered disrupted and can be evicted
regardless of whether the criteria in a PDB is met. 

: This means perspective running pods of a disrupted application might not get a chance to become healthy.

> By using this policy, cluster managers can easily evict misbehaving applications that are guarded by a PDB.

Author: I have put the same explanation in there, but I am not happy about the duplication.

  observedGeneration: 1
```

### Healthiness of a Pod

The current implementation considers healthy pods, as pods that have `.status.conditions` item with `type="Ready"` and `status="True"`.
These pods are tracked via `.status.currentHealthy` field in the PDB status.

## Unhealthy Pod Eviction Policy

{{< feature-state for_k8s_version="v1.26" state="alpha" >}}

{{< note >}}
In order to use this behavior, you must enable the `PDBUnhealthyPodEvictionPolicy`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
on the [API server](/docs/reference/command-line-tools-reference/kube-apiserver/).
{{< /note >}}

PodDisruptionBudget guarding an application ensures that `.status.currentHealthy` number of pods

> does not fall under `.status.desiredHealthy` number by not allowing eviction of healthy pods.

Author: ty, looks better

By using `.spec.unhealthyPodEvictionPolicy`, you can also define the criteria when unhealthy pods 
should be considered for eviction. The default behavior when no policy is specified corresponds 
to the `IfHealthyBudget` policy.

Policies:

`IfHealthyBudget`
: Running pods (`.status.phase="Running"`), but not yet healthy can be evicted only if the guarded application is not
disrupted (`.status.currentHealthy` is at least equal to `.status.desiredHealthy`).

: This policy ensures that running pods of an already disrupted application have the best chance to become healthy.
This has negative implications for draining nodes, which can be blocked by misbehaving applications that are guarded by a PDB.
More specifically applications with pods in `CrashLoopBackOff` state (due to a bug or misconfiguration),
or pods that are just failing to report the `Ready` condition.

`AlwaysAllow`
: Running pods (`.status.phase="Running"`), but not yet healthy are considered disrupted and can be evicted
regardless of whether the criteria in a PDB is met.

> : This means perspective running pods of a disrupted application might not get a chance to become healthy.

Author: We are trying to say that pods that are running but not healthy might take a long time before they get ready as they could have a bigger value of `.spec.minReadySeconds`.  So they are perspective in a sense that we expect them to become ready and not fail (as compared to the misbehaving pods). And with this policy if you are in an unstable environment (eg. when doing maintenance draining nodes, etc.) where lots of evictions are happening, these pods get evicted/killed without any chance to become ready. Which can cause the app to be disrupted for a longer period of time than desirable.

> : This means perspective running pods of a disrupted application might not get a chance to become healthy.

Reviewer 2: Oh, okay. So here's an edit based on what you just said. WDYT @atiratree ?
```suggestion
: Owing to a larger value of `spec.minReadySeconds`, prospective running pods of a disrupted application might not get a chance to become healthy. This, in turn, could lead to the pods being evicted or killed without any chance to become ready and cause the application to be disrupted for longer than desired.
```

> : This means perspective running pods of a disrupted application might not get a chance to become healthy.

Author: @divya-mohan0209 thanks, but I would rather not mention the `minReadySeconds`. I have mentioned that only to illustrate that this is where you would probably feel the most impact. But you could feel impact in any value, provided there is a sufficient amount of evictions.

> : This means perspective running pods of a disrupted application might not get a chance to become healthy.

Reviewer 2: Apart from that does the rest of it look okay to you?

> : This means perspective running pods of a disrupted application might not get a chance to become healthy.

Author: sorry, for the split conversation but there are other small issues

We do not need to mention killed since that is synonymous with evicted. And also we mean evicted in a reconcilling manner - not just a single time.
